packages/wp-now/src/pool.ts

```diff
@@ -0,0 +1,285 @@
+ * @param instance the php instance to clean up
+ * @param error the actual error that got us here
+ * @private
+ */
+const fatal = (pool: Pool, instance: instance, error: Error) => {
+	console.error(error);
+
+	if (pool.instanceInfo.has(instance)) {
+		const info = pool.instanceInfo.get(instance);
+		info.active = false;
+		pool.instanceInfo.delete(instance);
+	}
+
+	return pool.fatal(instance, error);
+};
+
+/**
+ * Find the next available idle instance.
+ * @private
+ */
+const getIdleInstance = (pool) => {
+	const sorted = [...pool.instanceInfo].sort(
+		(a, b) => a[1].requests - b[1].requests
+	);
+
+	for (const [instance, info] of sorted) {
+		if (pool.running.has(instance)) {
+			continue;
+		}
+
+		if (!info.active) {
+			continue;
+		}
+		return instance;
+	}
+
+	return false;
+};
+
+/**
+ * Maintains and refreshes a list of php instances
+ * such that each one will only be fed X number of requests
+ * before being discarded and replaced.
+ *
+ * Since we're dealing with a linear, "physical" memory array, as opposed to a
+ * virtual memory system afforded by most modern OSes, we're prone to things
+ * like memory fragmentation. In that situation, we could have the entire
+ * gigabyte empty except for a few sparse allocations. If no contiguous region
+ * of memory exists for the length requested, memory allocations will fail.
+ * This tends to happen when a new request attempts to initialize a heap
+ * structure but cannot find a contiguous 2mb chunk of memory.
+ *
+ * We can go as far as debugging PHP itself, and contributing the fix upstream.
+ * But even in this case we cannot guarantee that a third party extension will
+ * not introduce a leak sometime in the future. Therefore, we should have a
+ * solution robust to memory leaks that come from upstream code. I think that
+ * following the native strategy is the best way.
+ *
+ * https://www.php.net/manual/en/install.fpm.configuration.php#pm.max-requests
+ *
+ */
+export class Pool {
+	instanceInfo = new Map(); // php => PoolInfo
+
+	spawn: () => Promise<any>; // Async callback to create new instances.
+	fatal: (instance: instance, error: any) => any; // Async callback called on instance fatal errors.
+	reap:  (instance: instance) => void; // Async callback called on destroyed instances.
+	maxRequests: number; // Max requests to feed each instance
+	maxJobs: number; // Max number of instances to maintain at once.
+
+	resolvers = new Map(); // Inverted promises to notify async code of backlogged item processed.
+	running = new Set(); // Set of busy PHP instances.
+	backlog = []; // Set of request callbacks waiting to be run.
+
+	/**
+	 * Create a new pool.
+	 * @param options - {spawn, maxRequests, maxJobs}
+	 */
+	constructor({
+		maxRequests = 128,
+		maxJobs = 1,
+		spawn = async (): Promise<any> => {},
+		fatal = (instance: instance, request: request) => {},
+		reap = (instance: instance) => {},
+	} = {}) {
+		Object.defineProperties(this, {
```

Comment:
@dmsnell These props are now read-only.

Reply:
Which comment is this a reply to? I'd love to learn more about the context here.

Also, Are you sure these are read-only now? I don't see any mention of `defineProperties` marking props as read-only by default on that documentation page. Also, that would be a useful comment to include in the code with a rationale why it's important for them to be read-only.

Reply:
Aha, if the `value` is defined, but `writable` isn't, then `writable` defaults to false. That's not obvious from the code unless you know the JavaScript doc by heart, and even if you do it's still easy to miss. Let's make that code very obvious and explicitly state `writable: false`.

Reply:
https://github.com/WordPress/playground-tools/blob/f352bb84c1c67a45bbb27434c19b463f7a33b908/packages/wp-now/src/pool.ts#L160-L191

Reply:
@seanmorris I'd still like to learn more about the context here (as I asked in the comment above). What is this a reply to, and why do these properties need to be read only?

Reply:
@adamziel It doesn't really make sense to change these values, given the current pooling implementation. This all originally started based on the notation I was using to create private properties. Do you see a reason that calling code should be able to modify these values?